File: src/programmers/level_1/DFS/LV2_무인도_여행.kt
import java.util.*

class Solution {
    
    lateinit var isVisited:Array<BooleanArray>
    lateinit var maps:Array<String>
    
    fun solution(maps: Array<String>): IntArray {
        val answer=mutableListOf<Int>()
        this.maps=maps
        isVisited=Array(maps.size){BooleanArray(maps[0].length){false}}
        for(i in maps.indices){
            for(j in maps[0].indices){
                if(isVisited[i][j]||maps[i][j]=='X') continue
                isVisited[i][j]=true
                answer.add(dfs(i,j))
            }
        }
        return if(answer.isEmpty()) intArrayOf(-1) else answer.sorted().toIntArray()
    }
    
    val off=arrayOf(0 to 1, 0 to -1, 1 to 0, -1 to 0)
    fun dfs(x:Int,y:Int):Int{
        var res=(maps[x][y]-'0').toInt()
        for((ox,oy) in off){
            val nx=x+ox
            val ny=y+oy
            if(nx in 0 until maps.size && ny in 0 until maps[0].length){
                if(maps[nx][ny]!='X'&&!isVisited[nx][ny]){
                    isVisited[nx][ny]=true
                    res+=dfs(nx,ny)
                }
            }
        }
        return res
    }
}

File: src/algorithmProblems/programmers/level_1/DFS/LV2_무인도_여행.py
from collections import deque
def solution(maps):
    answer = []
    is_visited=[[False for _ in maps[0]] for _ in maps]
    
    for i,_ in enumerate(maps):
        for j,_ in enumerate(maps[0]):
            if(not is_visited[i][j] and maps[i][j]!='X'):
                answer.append(dfs(i,j,maps,is_visited))
    
    #print(is_visited)
    answer.sort()
    return [-1] if len(answer)==0 else answer

off=((0,1),(0,-1),(1,0),(-1,0))

def dfs(x,y,maps,is_visited):
    s=deque()
    c=int(maps[x][y])
    s.append((x,y))
    is_visited[x][y]=True
    
    while len(s)>0:
        cx,cy=s.pop()
        for ox,oy in off:
            nx=cx+ox
            ny=cy+oy
            if 0<=nx<len(maps) and 0<=ny<len(maps[0]) and not is_visited[nx][ny] and maps[nx][ny]!='X':
                is_visited[nx][ny]=True
                c+=int(maps[nx][ny])
                s.append((nx,ny))
    return c
        
        